File: app/MMarketx/js/src/data/flea/Frank.js
// Copyright 31-May-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

import Eflea from "./Eflea.js";

/**
    Ranking of Evaluated flea.
**/
export default class Frank {
  /* .
  _rc_ Frank: serial
    date : string
    efleas: !Array<!Eflea>
  */

  /*--*/
  /**
      @param {string} date
      @param {!Array<!Eflea>} efleas
  **/
  constructor (date, efleas) {

    /**
        @private
        @type {string}
    **/
    this._date = date;

    /**
        @private
        @type {!Array<!Eflea>}
    **/
    this._efleas = efleas;

  }

  /**
      @return {string}
  **/
  get date () {
    return this._date;
  }

  /**
      @return {!Array<!Eflea>}
  **/
  get efleas () {
    return this._efleas;
  }

  /**
      @return {!Array<?>}
  **/
  toJs () {
    return [
      this._date,
      this._efleas.map(e => e.toJs())
    ];
  }

  /**
      @param {!Array<?>} serial
      @return {!Frank}
  **/
  static fromJs (serial) {
    return new Frank(
      serial[0],
      serial[1].map(e =>
        Eflea.fromJs(e)
      )
    );
  }
  /*--*/
}

File: app/MultiMarket/go/data/flea/frank/frank.go
// Copyright 13-Sep-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Evaluate flea ranking entry.
package frank

import (
	"github.com/dedeme/MultiMarket/data/flea/eFlea"
	"github.com/dedeme/golib/json"
)

type T struct {
	date    string
	ranking []*eFlea.T
}

// Cretes a new 'frank'.
//    date   : Ranking date
//    ranking: Evaluated fleas list.
func New(date string, ranking []*eFlea.T) *T {
	return &T{date, ranking}
}

func (r *T) Date() string {
	return r.date
}

func (r *T) Ranking() []*eFlea.T {
	return r.ranking
}

func (r *T) ToJs() json.T {
	var rk []json.T
	for _, e := range r.ranking {
		rk = append(rk, e.ToJs())
	}
	return json.Wa([]json.T{
		json.Ws(r.date),
		json.Wa(rk),
	})
}

func FromJs(js json.T) *T {
	a := js.Ra()
	var rk []*eFlea.T
	for _, e := range a[1].Ra() {
		rk = append(rk, eFlea.FromJs(e))
	}
	return &T{
		a[0].Rs(),
		rk,
	}
}
